mat/cdense_arithmetic.go

```diff
@@ -0,0 +1,414 @@
+// Copyright ©2013 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/cblas128"
+)
+
+// Add adds a and b element-wise, placing the result in the receiver. Add
+// will panic if the two matrices do not have the same shape.
+func (m *CDense) Add(a, b CMatrix) {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br || ac != bc {
+		panic(ErrShape)
+	}
+
+	aU, aTrans, _ := untransposeCmplx(a)
+	bU, bTrans, _ := untransposeCmplx(b)
+
+	m.reuseAsNonZeroed(ar, ac)
+
+	if arm, ok := a.(*CDense); ok {
+		if brm, ok := b.(*CDense); ok {
+			amat, bmat := arm.mat, brm.mat
+			if m != aU {
+				m.checkOverlap(amat)
+			}
+			if m != bU {
+				m.checkOverlap(bmat)
+			}
+			for ja, jb, jm := 0, 0, 0; ja < ar*amat.Stride; ja, jb, jm = ja+amat.Stride, jb+bmat.Stride, jm+m.mat.Stride {
+				for i, v := range amat.Data[ja : ja+ac] {
+					m.mat.Data[i+jm] = v + bmat.Data[i+jb]
+				}
+			}
+			return
+		}
+	}
+
+	m.checkOverlapMatrix(aU)
+	m.checkOverlapMatrix(bU)
+	var restore func()
+	if aTrans && m == aU {
+		m, restore = m.isolatedWorkspace(aU)
+		defer restore()
+	} else if bTrans && m == bU {
+		m, restore = m.isolatedWorkspace(bU)
+		defer restore()
+	}
```

Comment:
These are not correct; if `a` is `m.T().H()` or `m.H().T()` then is is safe to write from `a` directly back to `m`. Equally, if `a` is `m.H()`, it's not. The approach here doesn't handle these cases, but the code in the comment below does; it's safe when the elements are transposed an even number of times (both H and T or neither).